### PR TITLE
Move Handler?.UpdateValue after PropertyChanged events

### LINKED ISSUE #5587: Shift Handler?.UpdateValue to happen after all PropertyChanged events have fired
Currently our `Handler?.UpdateValue` call fires during the top call here from `BindableObject`

image

This means that sometimes the `Handler` will have a premature vision of the world. The Handler update should fire after as many of the `PropertyChanged` events have settled as possible.

For example it leads to issues like this

https://github.com/dotnet/maui/pull/5586

The `Parent` on the `FormattedText` is set inside the `BP.OnPropertyChanged` which fires after the `Handler?.UpdateValue` fires. So anything the `BP.OnPropertyChanged` changes will now have to re-propagate.

### PR DESCRIPTION
**Pull request by antonfirsov**
### Description of Change

[Description updated two times]

This is a refactor to deal with #5587.

1. Introduce a private wrapper around `BindableObject.OnPropertyChanged` :

https://github.com/dotnet/maui/blob/c3e577694aebedd7adf3c37260e8b97ad45fbd45/src/Controls/src/Core/BindableObject.cs#L286-L291

And an internal property, `DelayHandlerUpdate` that indicates whether we are at non-zero depth at PropertyChanged invocations:

https://github.com/dotnet/maui/blob/c3e577694aebedd7adf3c37260e8b97ad45fbd45/src/Controls/src/Core/BindableObject.cs#L34-L37

2. Introduce a new virtual method `Element.UpdateHandler` to encapsulate a call to `Handler?.UpdateValue(propertyName)` and potential dependent Handler calls in subclasses.

3. In `Element.OnPropertyChanged`, do not call the handler update if we are in a call chain triggered by `SetProperty` or `ClearValue`. This is indicated by the new `DelayHandlerUpdate` property:

https://github.com/dotnet/maui/blob/c3e577694aebedd7adf3c37260e8b97ad45fbd45/src/Controls/src/Core/Element.cs#L385-L392

4. From BindableObject, when a property is updated through `SetProperty` or `ClearValue`, call the notifications in the following order:
- `OnPropertyChanged(suppressHandlerUpdate: true, name)`
- Fire `property.PropertyChanged`
- Update all the delayed handlers, including dependent ones. This is done by `Element.RunDelayedHandlerUpdates` which also has a protection against re-entrant calls:

https://github.com/dotnet/maui/blob/c3e577694aebedd7adf3c37260e8b97ad45fbd45/src/Controls/src/Core/Element.cs#L419-L436

Manual calls to `OnPropertyChanged(name)` do not delay the handler update.

/cc @PureWeen 

### Issues Fixed

Fixes #5587

### CODE REVIEW

File: src/Compatibility/ControlGallery/src/Issues.Shared/Bugzilla31114.cs [hunk 1]
```diff
@@ -325,9 +325,9 @@ void Init(bool fast)
 			}
 
 
-			protected override void OnPropertyChanged(string propertyName = null)
```
Comment (PureWeen) [inline src/Compatibility/ControlGallery/src/Issues.Shared/Bugzilla31114.cs:328]: Can we make anything that's modifying a public API internal or `private protected` for now? Then we can look at making something public for NET7?

Comment (antonfirsov) [reply src/Compatibility/ControlGallery/src/Issues.Shared/Bugzilla31114.cs:328]: The problem is that we can't have two `OnPropertyChanged` methods as I originally wanted. Should we remove `OnPropertyChanged` from the public API then?

**Comment by antonfirsov** [comment]
@PureWeen @mattleibow I implemented a logic to deal with nested property changes, updated the description once again.

**Comment by PureWeen** [comment]
Closing this for now

We'll revisit this PR once the associated issue bubbles up to our ongoing board.

**Pull request closed without merging**